File: internal/registry/registry.go
package registry

import (
	"errors"
	"fmt"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"
)

const registryRelativePath = "woon-core/registry/repositories.yaml"

type Repository struct {
	Remote    string `yaml:"remote"`
	Directory string `yaml:"directory"`
	Role      string `yaml:"role,omitempty"`
	Output    bool   `yaml:"output,omitempty"`
}

type Registry struct {
	Version      int                   `yaml:"version"`
	Repositories map[string]Repository `yaml:"repositories"`
}

type SyncResult struct {
	Cloned   int
	Existing int
}

func Load(root string) (Registry, error) {
	path := filepath.Join(root, filepath.FromSlash(registryRelativePath))
	data, err := os.ReadFile(path)
	if err != nil {
		return Registry{}, fmt.Errorf("read registry %s: %w", path, err)
	}
	var reg Registry
	if err := yaml.Unmarshal(data, &reg); err != nil {
		return Registry{}, fmt.Errorf("parse registry: %w", err)
	}
	if err := reg.Validate(); err != nil {
		return Registry{}, err
	}
	return reg, nil
}

func (r Registry) Validate() error {
	if r.Version != 1 {
		return fmt.Errorf("unsupported registry version %d", r.Version)
	}
	seen := map[string]string{}
	for id, repo := range r.Repositories {
		if id == "" || repo.Directory == "" || repo.Remote == "" {
			return fmt.Errorf("repository %q requires remote and directory", id)
		}
		if filepath.IsAbs(repo.Directory) || filepath.Clean(repo.Directory) != repo.Directory || strings.Contains(repo.Directory, "..") {
			return fmt.Errorf("repository %q has unsafe directory %q", id, repo.Directory)
		}
		if previous, ok := seen[repo.Directory]; ok {
			return fmt.Errorf("repositories %q and %q share directory %q", previous, id, repo.Directory)
		}
		seen[repo.Directory] = id
		parsed, err := url.Parse(repo.Remote)
		if err != nil || parsed.Scheme != "https" || parsed.Host != "github.com" {
			return fmt.Errorf("repository %q has unsupported remote %q", id, repo.Remote)
		}
	}
	return nil
}

func (r Registry) Resolve(root, reference string) (string, error) {
	id, relative, err := parseReference(reference)
	if err != nil {
		return "", err
	}
	repo, ok := r.Repositories[id]
	if !ok {
		return "", fmt.Errorf("unknown repository %q", id)
	}
	base := filepath.Join(root, repo.Directory)
	resolved := filepath.Join(base, filepath.FromSlash(relative))
	if resolved != base && !strings.HasPrefix(resolved, base+string(filepath.Separator)) {
		return "", fmt.Errorf("reference escapes repository %q", id)
	}
	return resolved, nil
}

func (r Registry) Missing(root string) []string {
	var missing []string
	for id, repo := range r.Repositories {
		if _, err := os.Stat(filepath.Join(root, repo.Directory)); errors.Is(err, os.ErrNotExist) {
			missing = append(missing, id)
		}
	}
	sort.Strings(missing)
	return missing
}

func (r Registry) Sync(root string) (SyncResult, error) {
	var result SyncResult
	ids := make([]string, 0, len(r.Repositories))
	for id := range r.Repositories {
		ids = append(ids, id)
	}
	sort.Strings(ids)
	for _, id := range ids {
		repo := r.Repositories[id]
		target := filepath.Join(root, repo.Directory)
		if _, err := os.Stat(target); err == nil {
			if _, err := os.Stat(filepath.Join(target, ".git")); err != nil {
				return result, fmt.Errorf("%s exists but is not a Git checkout", target)
			}
			result.Existing++
			continue
		} else if !errors.Is(err, os.ErrNotExist) {
			return result, err
		}
		cmd := exec.Command("git", "clone", "--", repo.Remote, target)
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
		if err := cmd.Run(); err != nil {
			return result, fmt.Errorf("clone %s: %w", id, err)
		}
		result.Cloned++
	}
	return result, nil
}

func parseReference(reference string) (string, string, error) {
	if !strings.HasPrefix(reference, "repo://") {
		if strings.Contains(reference, "/") || reference == "" {
			return "", "", fmt.Errorf("invalid repository ID %q", reference)
		}
		return reference, "", nil
	}
	rest := strings.TrimPrefix(reference, "repo://")
	parts := strings.SplitN(rest, "/", 2)
	if parts[0] == "" {
		return "", "", fmt.Errorf("repo URI requires an ID")
	}
	relative := ""
	if len(parts) == 2 {
		relative = filepath.ToSlash(filepath.Clean(parts[1]))
		if relative == "." {
			relative = ""
		}
		if relative == ".." || strings.HasPrefix(relative, "../") {
			return "", "", fmt.Errorf("repo URI may not escape its repository")
		}
	}
	return parts[0], relative, nil
}

File: src/woon_core/registry.py
"""Repository registry validation and path resolution."""

from __future__ import annotations

import subprocess
from dataclasses import dataclass
from pathlib import Path, PurePosixPath
from urllib.parse import urlparse

from woon_core.errors import WoonError
from woon_core.io import load_yaml

REGISTRY_RELATIVE_PATH = Path("woon-core/registry/repositories.yaml")


@dataclass(frozen=True, slots=True)
class Repository:
    remote: str
    directory: str
    role: str = ""
    output: bool = False


@dataclass(frozen=True, slots=True)
class SyncResult:
    cloned: int
    existing: int


@dataclass(frozen=True, slots=True)
class Registry:
    version: int
    repositories: dict[str, Repository]

    @classmethod
    def load(cls, root: Path) -> Registry:
        path = root / REGISTRY_RELATIVE_PATH
        raw = load_yaml(path)
        raw_repositories = raw.get("repositories")
        if not isinstance(raw_repositories, dict):
            raise WoonError("registry requires a repositories mapping")
        repositories: dict[str, Repository] = {}
        for identifier, item in raw_repositories.items():
            if not isinstance(identifier, str) or not isinstance(item, dict):
                raise WoonError("registry repository entries must be mappings")
            repositories[identifier] = Repository(
                remote=str(item.get("remote", "")),
                directory=str(item.get("directory", "")),
                role=str(item.get("role", "")),
                output=bool(item.get("output", False)),
            )
        registry = cls(version=int(raw.get("version", 0)), repositories=repositories)
        registry.validate()
        return registry

    def validate(self) -> None:
        if self.version != 1:
            raise WoonError(f"unsupported registry version {self.version}")
        seen_directories: dict[str, str] = {}
        for identifier, repository in self.repositories.items():
            if not identifier or not repository.directory or not repository.remote:
                raise WoonError(f"repository {identifier!r} requires remote and directory")
            directory = PurePosixPath(repository.directory)
            if (
                directory.is_absolute()
                or ".." in directory.parts
                or str(directory) != repository.directory
            ):
                raise WoonError(
                    f"repository {identifier!r} has unsafe directory {repository.directory!r}"
                )
            if previous := seen_directories.get(repository.directory):
                raise WoonError(
                    f"repositories {previous!r} and {identifier!r} share directory "
                    f"{repository.directory!r}"
                )
            seen_directories[repository.directory] = identifier
            parsed = urlparse(repository.remote)
            if parsed.scheme != "https" or parsed.hostname != "github.com":
                raise WoonError(
                    f"repository {identifier!r} has unsupported remote {repository.remote!r}"
                )

    def resolve(self, root: Path, reference: str) -> Path:
        identifier, relative = _parse_reference(reference)
        try:
            repository = self.repositories[identifier]
        except KeyError as error:
            raise WoonError(f"unknown repository {identifier!r}") from error
        base = (root / repository.directory).resolve(strict=False)
        resolved = (base / relative).resolve(strict=False)
        if not resolved.is_relative_to(base):
            raise WoonError(f"reference escapes repository {identifier!r}")
        return resolved

    def missing(self, root: Path) -> list[str]:
        return sorted(
            identifier
            for identifier, repository in self.repositories.items()
            if not (root / repository.directory).exists()
        )

    def sync(self, root: Path) -> SyncResult:
        cloned = 0
        existing = 0
        for identifier in sorted(self.repositories):
            repository = self.repositories[identifier]
            target = root / repository.directory
            if target.exists():
                if not (target / ".git").exists():
                    raise WoonError(f"{target} exists but is not a Git checkout")
                existing += 1
                continue
            try:
                subprocess.run(["git", "clone", "--", repository.remote, str(target)], check=True)
            except subprocess.CalledProcessError as error:
                raise WoonError(f"clone {identifier}: {error}") from error
            cloned += 1
        return SyncResult(cloned=cloned, existing=existing)


def _parse_reference(reference: str) -> tuple[str, Path]:
    if not reference.startswith("repo://"):
        if not reference or "/" in reference:
            raise WoonError(f"invalid repository ID {reference!r}")
        return reference, Path()
    rest = reference.removeprefix("repo://")
    identifier, separator, raw_relative = rest.partition("/")
    if not identifier:
        raise WoonError("repo URI requires an ID")
    relative = PurePosixPath(raw_relative) if separator else PurePosixPath()
    if relative.is_absolute() or ".." in relative.parts:
        raise WoonError("repo URI may not escape its repository")
    return identifier, Path(*relative.parts)
